// @officer/embedded-fonts/dist/wasm-decoder.js
import { DEFAULT_EMBEDDED_FONT_DECODE_POLICY } from "./types.js";
const SUPPORTED_EOT_CONTENT_TYPES = new Set([
    "application/x-fontdata",
    "application/vnd.ms-fontobject",
]);
const textDecoder = new TextDecoder();
export async function createEmbeddedFontWasmDecoder(source) {
    const wasm = await instantiateEmbeddedFontDecoderWasm(source);
    return (input) => decodeEmbeddedFontWithWasm(input, wasm);
}
export async function decodeEmbeddedFontWithWasm(input, wasm) {
    const policy = resolvePolicy(input.policy);
    if (!SUPPORTED_EOT_CONTENT_TYPES.has(input.contentType)) {
        return statusResult(input, "unsupported_format", [
            {
                code: "embedded_font_unsupported_content_type",
                severity: "warning",
                message: `Unsupported embedded font content type "${input.contentType}".`,
                family: input.sourceFamily,
                relationshipId: input.relationshipId,
            },
        ]);
    }
    if (input.data.byteLength > policy.maxInputBytes) {
        return statusResult(input, "decode_failed", [
            {
                code: "embedded_font_input_too_large",
                severity: "error",
                message: `Embedded font input has ${input.data.byteLength} bytes, exceeding limit ${policy.maxInputBytes}.`,
                family: input.sourceFamily,
                relationshipId: input.relationshipId,
            },
        ]);
    }
    let frame;
    try {
        frame = decodeWasmFrame(wasm, input.data, policy);
    }
    catch (error) {
        return statusResult(input, "decode_failed", [
            {
                code: "embedded_font_wasm_decode_failed",
                severity: "error",
                message: error instanceof Error
                    ? error.message
                    : `Embedded font WASM decode failed: ${String(error)}`,
                family: input.sourceFamily,
                relationshipId: input.relationshipId,
            },
        ]);
    }
    const { header, decodedFont } = parseWasmFrame(frame);
    if (!header.ok) {
        return statusResult(input, header.status ?? "decode_failed", [
            {
                code: "embedded_font_decode_failed",
                severity: "error",
                message: header.message,
                family: input.sourceFamily,
                relationshipId: input.relationshipId,
            },
        ]);
    }
    const diagnostics = header.warnings.map((warning) => warningDiagnostic(warning, input));
    if (!header.policy.renderRegistrationAllowed &&
        !policy.allowRestrictedRendering) {
        return statusResult(input, "restricted", [
            ...diagnostics,
            {
                code: "embedded_font_restricted_by_policy",
                severity: "warning",
                message: `Embedded font permission "${header.policy.permission}" is not allowed for render registration.`,
                family: input.sourceFamily,
                relationshipId: input.relationshipId,
            },
        ], header);
    }
    if (header.metadata.eot?.subsetted === true) {
        return statusResult(input, "subsetted", [
            ...diagnostics,
            {
                code: "embedded_font_subset_missing_glyph",
                severity: "warning",
                message: "Embedded font is subsetted; glyph coverage checks are not available yet.",
                family: input.sourceFamily,
                relationshipId: input.relationshipId,
            },
        ], header);
    }
    if (decodedFont.byteLength > policy.maxOutputBytes) {
        return statusResult(input, "decode_failed", [
            ...diagnostics,
            {
                code: "embedded_font_output_too_large",
                severity: "error",
                message: `Decoded embedded font has ${decodedFont.byteLength} bytes, exceeding limit ${policy.maxOutputBytes}.`,
                family: input.sourceFamily,
                relationshipId: input.relationshipId,
            },
        ], header);
    }
    return {
        status: "decoded",
        sourceFamily: input.sourceFamily,
        sourceFaceType: input.sourceFaceType,
        contentType: input.contentType,
        relationshipId: input.relationshipId,
        inputSha256: input.inputSha256,
        outputSha256: await sha256WithWebCrypto(decodedFont),
        decodedFont,
        outputFormat: header.outputFormat,
        metadata: header.metadata,
        registration: header.registration,
        diagnostics,
    };
}
async function instantiateEmbeddedFontDecoderWasm(source) {
    const instance = await instantiateWasm(source);
    return readWasmExports(instance);
}
async function instantiateWasm(source) {
    if (source instanceof WebAssembly.Module) {
        return WebAssembly.instantiate(source, {});
    }
    const module = await WebAssembly.compile(source);
    return WebAssembly.instantiate(module, {});
}
function readWasmExports(instance) {
    const exports = instance.exports;
    const memory = exports["memory"];
    if (!(memory instanceof WebAssembly.Memory)) {
        throw new Error("Embedded font decoder WASM did not export memory.");
    }
    const alloc = requiredWasmFunction(exports, "embedded_font_decoder_alloc");
    const free = requiredWasmFunction(exports, "embedded_font_decoder_free");
    const decodeWithLimits = requiredWasmFunction(exports, "embedded_font_decoder_decode_with_limits");
    const resultPtr = requiredWasmFunction(exports, "embedded_font_decoder_result_ptr");
    const resultLen = requiredWasmFunction(exports, "embedded_font_decoder_result_len");
    const clearResult = requiredWasmFunction(exports, "embedded_font_decoder_clear_result");
    return {
        memory,
        embedded_font_decoder_alloc: alloc,
        embedded_font_decoder_free: free,
        embedded_font_decoder_decode_with_limits: decodeWithLimits,
        embedded_font_decoder_result_ptr: resultPtr,
        embedded_font_decoder_result_len: resultLen,
        embedded_font_decoder_clear_result: clearResult,
    };
}
function requiredWasmFunction(exports, name) {
    const value = exports[name];
    if (typeof value !== "function") {
        throw new Error(`Embedded font decoder WASM missing export ${name}.`);
    }
    return value;
}
function decodeWasmFrame(wasm, data, policy) {
    const inputPtr = wasm.embedded_font_decoder_alloc(data.byteLength);
    try {
        new Uint8Array(wasm.memory.buffer, inputPtr, data.byteLength).set(data);
        wasm.embedded_font_decoder_decode_with_limits(inputPtr, data.byteLength, policy.maxInputBytes, policy.maxOutputBytes);
        const resultPtr = wasm.embedded_font_decoder_result_ptr();
        const resultLen = wasm.embedded_font_decoder_result_len();
        if (resultPtr === 0 || resultLen === 0) {
            throw new Error("Embedded font decoder WASM returned an empty result.");
        }
        return new Uint8Array(new Uint8Array(wasm.memory.buffer, resultPtr, resultLen));
    }
    finally {
        wasm.embedded_font_decoder_free(inputPtr, data.byteLength);
        wasm.embedded_font_decoder_clear_result();
    }
}
function parseWasmFrame(frame) {
    if (frame.byteLength < 4) {
        throw new Error("Embedded font decoder WASM result frame is too small.");
    }
    const jsonLength = new DataView(frame.buffer, frame.byteOffset, 4).getUint32(0, true);
    const decodedOffset = 4 + jsonLength;
    if (decodedOffset > frame.byteLength) {
        throw new Error("Embedded font decoder WASM result frame is truncated.");
    }
    const headerText = textDecoder.decode(frame.subarray(4, decodedOffset));
    const parsed = JSON.parse(headerText);
    return {
        header: parseWasmDecodeHeader(parsed),
        decodedFont: new Uint8Array(frame.subarray(decodedOffset)),
    };
}
function parseWasmDecodeHeader(value) {
    const header = objectValue(value, "WASM decode header");
    const ok = header["ok"];
    if (ok === true) {
        return {
            ok: true,
            outputFormat: outputFormatValue(header["outputFormat"]),
            metadata: metadataValue(header["metadata"]),
            registration: registrationValue(header["registration"]),
            policy: policyValue(header["policy"]),
            warnings: warningsValue(header["warnings"]),
        };
    }
    if (ok === false) {
        return {
            ok: false,
            status: typeof header["status"] === "string"
                ? decodeStatusValue(header["status"])
                : undefined,
            message: stringValue(header["message"], "WASM decode error message"),
        };
    }
    throw new Error("Embedded font decoder WASM header missing ok boolean.");
}
function metadataValue(value) {
    const metadata = objectValue(value, "WASM metadata");
    return {
        eot: metadata["eot"] === undefined
            ? undefined
            : eotMetadataValue(metadata["eot"]),
        sfnt: metadata["sfnt"] === undefined
            ? undefined
            : sfntMetadataValue(metadata["sfnt"]),
    };
}
function eotMetadataValue(value) {
    const eot = objectValue(value, "WASM EOT metadata");
    return {
        version: eot["version"] === undefined
            ? undefined
            : eotVersionValue(eot["version"]),
        flags: optionalNumberValue(eot["flags"], "EOT flags"),
        subsetted: optionalBooleanValue(eot["subsetted"], "EOT subsetted"),
        compressed: optionalBooleanValue(eot["compressed"], "EOT compressed"),
        xorEncrypted: optionalBooleanValue(eot["xorEncrypted"], "EOT xorEncrypted"),
        permissions: optionalNumberValue(eot["permissions"], "EOT permissions"),
        familyName: optionalStringValue(eot["familyName"], "EOT familyName"),
        styleName: optionalStringValue(eot["styleName"], "EOT styleName"),
        fullName: optionalStringValue(eot["fullName"], "EOT fullName"),
    };
}
function sfntMetadataValue(value) {
    const sfnt = objectValue(value, "WASM SFNT metadata");
    return {
        version: sfnt["version"] === undefined
            ? undefined
            : sfntVersionValue(sfnt["version"]),
        tableTags: sfnt["tableTags"] === undefined
            ? undefined
            : stringArrayValue(sfnt["tableTags"], "SFNT tableTags"),
    };
}
function registrationValue(value) {
    const registration = objectValue(value, "WASM registration metadata");
    const cmap = objectValue(registration["cmap"], "WASM cmap metadata");
    return {
        family: optionalStringValue(registration["family"], "registration family"),
        subfamily: optionalStringValue(registration["subfamily"], "registration subfamily"),
        fullName: optionalStringValue(registration["fullName"], "registration fullName"),
        weight: optionalNumberValue(registration["weight"], "registration weight"),
        italic: optionalBooleanValue(registration["italic"], "registration italic"),
        cmap: {
            subtableCount: numberValue(cmap["subtableCount"], "cmap subtableCount"),
            selectedPlatformId: numberValue(cmap["selectedPlatformId"], "cmap selectedPlatformId"),
            selectedEncodingId: numberValue(cmap["selectedEncodingId"], "cmap selectedEncodingId"),
            selectedFormat: numberValue(cmap["selectedFormat"], "cmap selectedFormat"),
            codepointCount: numberValue(cmap["codepointCount"], "cmap codepointCount"),
            firstCodepoint: optionalNumberValue(cmap["firstCodepoint"], "cmap firstCodepoint"),
            lastCodepoint: optionalNumberValue(cmap["lastCodepoint"], "cmap lastCodepoint"),
            hasBasicLatin: booleanValue(cmap["hasBasicLatin"], "cmap hasBasicLatin"),
        },
    };
}
function policyValue(value) {
    const policy = objectValue(value, "WASM embedded font policy");
    return {
        permission: stringValue(policy["permission"], "policy permission"),
        renderRegistrationAllowed: booleanValue(policy["renderRegistrationAllowed"], "policy renderRegistrationAllowed"),
        editableReuseAllowed: booleanValue(policy["editableReuseAllowed"], "policy editableReuseAllowed"),
        reexportAllowed: booleanValue(policy["reexportAllowed"], "policy reexportAllowed"),
        rawBytesRoundTripAllowed: booleanValue(policy["rawBytesRoundTripAllowed"], "policy rawBytesRoundTripAllowed"),
        noSubsetting: booleanValue(policy["noSubsetting"], "policy noSubsetting"),
        bitmapEmbeddingOnly: booleanValue(policy["bitmapEmbeddingOnly"], "policy bitmapEmbeddingOnly"),
    };
}
function warningsValue(value) {
    if (!Array.isArray(value)) {
        throw new Error("WASM warnings must be an array.");
    }
    return value.map((entry) => {
        const warning = objectValue(entry, "WASM warning");
        const kind = stringValue(warning["kind"], "warning kind");
        if (kind === "name_mismatch") {
            return {
                kind,
                field: stringValue(warning["field"], "warning field"),
                eot: stringValue(warning["eot"], "warning eot"),
                sfnt: stringValue(warning["sfnt"], "warning sfnt"),
            };
        }
        if (kind === "missing_decoded_name") {
            return {
                kind,
                field: stringValue(warning["field"], "warning field"),
                eot: stringValue(warning["eot"], "warning eot"),
            };
        }
        throw new Error(`Unknown WASM warning kind ${kind}.`);
    });
}
function statusResult(input, status, diagnostics, header) {
    return {
        status,
        sourceFamily: input.sourceFamily,
        sourceFaceType: input.sourceFaceType,
        contentType: input.contentType,
        relationshipId: input.relationshipId,
        inputSha256: input.inputSha256,
        metadata: header?.metadata ?? {},
        registration: header?.registration,
        diagnostics,
    };
}
function warningDiagnostic(warning, input) {
    if (warning.kind === "name_mismatch") {
        return {
            code: "embedded_font_name_mismatch",
            severity: "warning",
            message: `EOT ${warning.field} "${warning.eot}" does not match decoded SFNT name "${warning.sfnt}".`,
            family: input.sourceFamily,
            relationshipId: input.relationshipId,
        };
    }
    return {
        code: "embedded_font_missing_decoded_name",
        severity: "warning",
        message: `EOT ${warning.field} "${warning.eot}" is missing from decoded SFNT name metadata.`,
        family: input.sourceFamily,
        relationshipId: input.relationshipId,
    };
}
function resolvePolicy(policy) {
    return {
        ...DEFAULT_EMBEDDED_FONT_DECODE_POLICY,
        ...policy,
    };
}
function objectValue(value, field) {
    if (typeof value !== "object" || value === null || Array.isArray(value)) {
        throw new Error(`${field} must be an object.`);
    }
    return value;
}
function stringValue(value, field) {
    if (typeof value !== "string") {
        throw new Error(`${field} must be a string.`);
    }
    return value;
}
function optionalStringValue(value, field) {
    return value === undefined ? undefined : stringValue(value, field);
}
function numberValue(value, field) {
    if (typeof value !== "number" || !Number.isFinite(value)) {
        throw new Error(`${field} must be a finite number.`);
    }
    return value;
}
function optionalNumberValue(value, field) {
    return value === undefined ? undefined : numberValue(value, field);
}
function booleanValue(value, field) {
    if (typeof value !== "boolean") {
        throw new Error(`${field} must be a boolean.`);
    }
    return value;
}
function optionalBooleanValue(value, field) {
    return value === undefined ? undefined : booleanValue(value, field);
}
function stringArrayValue(value, field) {
    if (!Array.isArray(value)) {
        throw new Error(`${field} must be an array.`);
    }
    return value.map((entry) => stringValue(entry, field));
}
function outputFormatValue(value) {
    const format = stringValue(value, "output format");
    switch (format) {
        case "otf":
        case "ttc":
        case "ttf":
        case "woff":
        case "woff2":
            return format;
    }
    throw new Error(`Unsupported output format ${format}.`);
}
function decodeStatusValue(value) {
    switch (value) {
        case "decoded":
        case "unsupported_format":
        case "restricted":
        case "subsetted":
        case "decode_failed":
            return value;
    }
    throw new Error(`Unsupported decode status ${value}.`);
}
function eotVersionValue(value) {
    switch (numberValue(value, "EOT version")) {
        case 1:
            return 1;
        case 2:
            return 2;
        case 3:
            return 3;
        default:
            throw new Error(`Unsupported EOT version ${String(value)}.`);
    }
}
function sfntVersionValue(value) {
    const version = stringValue(value, "SFNT version");
    switch (version) {
        case "AppleTrueType":
        case "OpenTypeCff":
        case "PostScriptType1":
        case "TrueType":
            return version;
    }
    throw new Error(`Unsupported SFNT version ${version}.`);
}
async function sha256WithWebCrypto(data) {
    const subtle = globalThis.crypto?.subtle;
    if (!subtle) {
        return undefined;
    }
    const digest = await subtle.digest("SHA-256", data);
    return Array.from(new Uint8Array(digest))
        .map((byte) => byte.toString(16).padStart(2, "0"))
        .join("");
}
